[NVIDIA TF] Support building against CUDA 12.0

nluehr opened the pull request:
This PR updates TensorFlow to build against CUDA 12.0. Most changes are minor with the exception of the replacing the csrGemmV2 APIs with SpGEMM, since the former was removed from cusparse 12.0.

Attn: @hawkinsp

nluehr committed:
Migrate remaining calls to cusparseCsrmvEx to cusparseSpMV for CUDA 12

nluehr committed:
Replace csrGemmV2 with calls to SpGEMM APIs when compiling for CUDA 12.

nluehr committed:
Switch from CUSPARSE_CSR2CSC_ALG2 to CUSPARSE_CSR2CSC_ALG1 for CUDA >= 12

nluehr committed:
Update algorithm enum sparse mat_mul_op for CUDA 12

As of CUDA 12, CUSPARSE_MM_ALG_DEFAULT is replaced by CUSPARSE_SPMM_ALG_DEFAULT.

nluehr committed:
Use major version for cupti lib with CUDA>=12

nluehr committed:
Update CUDA stub libraries for CUDA 12

nluehr committed:
Bump NCCL version to 2.16.2 to support CUDA 12 and NVIDIA Hopper GPUs.

nluehr committed:
Update cudaGraph APIs for CUDA 12.

cantonios commented on tensorflow/core/kernels/sparse/sparse_mat_mul_op.cc:
```diff
@@ -417,6 +411,136 @@ class CSRSparseMatMulGPUOp : public OpKernel {
+    // (along with the gemmDesc) contains intermediate products needed for
+    // writing the final output. Because the final outputs cannot be allocated
+    // until the NNZ is computed for the entire batch, we need to keep
+    // batch_size buffer2 temoraries.
```
nit: spelling "temporaries"

cantonios replied:
Isn't this a bit unfortunate?  According to the docs, 

> the first invocation of `cusparseSpGEMM_compute` provides an upper bound of the memory required for the computation that is generally several times larger of the actual memory used

So we're potentially allocating many larger-than-necessary buffers, one for each element in the batch.  Would it make sense to allocate a single `buffer2` (that potentially grows as needed), and a potentially growing buffer to store the final columns/values?

nluehr replied:
There are a few options to optimize. I wasn't sure which way to go, so I took the simplest option, which also has the virtue of being the fastest (assuming sufficient memory is available).

If we go with a growing result buffer, we can either reallocate for every element of the batch or estimate the output size based on the NNZ counts observed so far within the batch. A few options ...

* Reallocate at every iteration (e.g. assume all remaining gemms require size zero).
* Reallocate by multiples of two to fit the size required at each iteration
* Assume each remaining GEMM requires the same size as the largest NNZ observed so far in batch.
* Assume each remaining GEMM requires Average NNZ observed so far in batch.
* Either of the above, but add a statistical margin, like an extra 10% or some multiple of the stdev, or space for batch_size+1 outputs.

The "max" option could overestimate the required memory by a lot if one of the gemms is significantly more dense than all the others (worst case by a factor of batch size). Of course, this applies to the "average" case as well, if the outlier happens to be the first gemm. In strange corner cases, this could be larger than the "several times" over allocation in the current implementation, and could persist to downstream ops, unless the output is always resized to an exact fit. Only the "dumbest" first option will never over allocate memory. The second option would bound the over-allocation by a factor of two.

I'm happy to implement any of these. @cantonios, which do you think is best?

cantonios replied:
I was thinking either the 2nd or the 4th one.  I had been debating between those two options myself before your PR when we found out CUDA 12 dropped the old API.  The final output should be resized if it is overly large.

I understand all those tradeoffs, so I'm also on the fence about the best approach.  In the current implementation, the intermediate buffers are all still live when we allocate the result buffers, so the total memory allocated will potentially be considerably more than 2x the final required result size.  It kind of depends how much "several times larger" the intermediate buffers actually are in practice.

I leave it to you to decide what's best here.  I'm okay with the current solution, but thought I would flag my concerns since I had been thinking about it myself over the last week or so.

nluehr replied:
I went with using separate intermediate buffers for each gemm in the batch and then finishing up with a single concat across the outputs.
@cantonios PTAL.

gbaned commented:
Hi @cantonios Can you please review this PR ? Thank you!

cantonios commented:
> Hi @cantonios Can you please review this PR ? Thank you!

I added my review, and there was no response.

nluehr committed:
Reduce memory overheads in sparse-sparse matmul.

Memory reduction comes at the cost of an additional device-side copy
to concat the gemm results across the batch.

cantonios reviewed (approved):
Thanks!

nluehr committed:
Guard cuda_blas_utils include to fix ROCM build.

nluehr committed:
Load cupti dso using correct version.

Fixes issue where minor version was incorrectly included in dso
name with cuda 12.

nluehr commented:
Added a fix to use cupti_version (major version only for cuda 12 and later) rather than cuda_version (major.minor version) to load the libcupti DSO.

nluehr commented:
@reedwm is this blocked? Anything I can do on my side to help?

reedwm commented:
It was blocked but now it simply needs to be approved internally. It probably will be merged Monday.

luckeyca commented:
does this cover CUDA 12 in WSL2 as I ran into issue with tensorflow right after pip install. verification command was looking for CUDA 11 library instead of 12 installed on the WSL2 instance. https://github.com/tensorflow/tensorflow/issues/59413

nluehr commented:
This PR enables building TF from source against CUDA 12. The nightly and release builds available from PyPI continue to be built at present against CUDA 11.8.

yangtj207 commented:
Is there a plan to provide docker images that work with CUDA 12.0? Thank you.

aminalaee commented:
Can anyone share the plan for when this will be released?

alanwilter commented:
I tried `pip install tensorflow==2.12.rc1` and still not working with CUDA 12. Which release should see TF working with CUDA 12?

nluehr commented:
@alanwilter Presently you need to build TensorFlow from source to use it with CUDA 12.x. Either the master or the r2.12 release branches will build against CUDA 12.
Alternatively, you could use NVIDIA's NGC containers [here](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorflow) which are pre-built against CUDA 12.

ddelange commented:
> This PR enables building TF from source against CUDA 12. The nightly and release builds available from PyPI continue to be built at present against CUDA 11.8.

Is there a timeline for building the official wheels against CUDA 12.x?

Talador12 commented:
> This PR enables building TF from source against CUDA 12. The nightly and release builds available from PyPI continue to be built at present against CUDA 11.8.

These 12.x packages need to be built on PyPI. CUDA 11.x is deprecated in some systems

nluehr commented:
@Talador12 can you provide more information about where CUDA 11.8 is deprecated?

Talador12 commented:
It could be a few reasons, but Fedramp compliance and using current Debian versions.

I was surprised that Tensorflow did not have prebuilt CUDA 12 support on PyPi.

Talador12 commented:
Is there an update on this issue? There is still a need for a CUDA 12 build of tensorflow on Pypi

Talador12 commented:
Could we re-open this issue? This has not been resolved yet

reedwm commented:
This is a PR that has been merged, not an issue, so it cannot be reopened.

We have not yet released pip packages with CUDA 12 support, but are working on this. Feel free to file a new GitHub issue to have CUDA 12 pip packages (please CC me on the issue if you file it).

Talador12 commented:
Apologies - I thought the python package would be built using the merged code in this pull request. I created a separate issue for the python package at #60943
